os/net/ipv6/multicast/mpl.c

```diff
@@ -0,0 +1,1807 @@
+#define LOG_SEED(level, seed_id) do { \
+    if(level <= (LOG_LEVEL)) { \
+      LOG_OUTPUT("0x%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx%.2hx", \
+                 seed_id.id[15], seed_id.id[14], seed_id.id[13], seed_id.id[12], \
+                 seed_id.id[11], seed_id.id[10], seed_id.id[9], seed_id.id[8], \
+                 seed_id.id[7], seed_id.id[6], seed_id.id[5], seed_id.id[4], \
+                 seed_id.id[3], seed_id.id[2], seed_id.id[1], seed_id.id[0]); \
+    } \
+} while(0);
+/* Macros to print seed id in logs */
+#define LOG_INFO_SEED(...) LOG_SEED(LOG_LEVEL_INFO, __VA_ARGS__)
+#define LOG_WARN_SEED(...) LOG_SEED(LOG_LEVEL_WARN, __VA_ARGS__)
+#define LOG_ERR_SEED(...) LOG_SEED(LOG_LEVEL_ERR, __VA_ARGS__)
+#define LOG_DBG_SEED(...) LOG_SEED(LOG_LEVEL_DBG, __VA_ARGS__)
+
+/* MPL Seed IDs can be either 16 bits, 64 bits, or 128 bits. If they are 128
+ *  bits then the IPV6 Source address may also be used as the seed ID.
+ *  These are always represented in contiki as a 128 bit number in the type
+ *  seed_id_t. The functions below convert a seed id of various lengths to
+ *  this 128 bit representation.
+ */
+/**
+ * \brief Set the seed id to a 16 bit constant
+ * dst: seed_id_t to set to the constant
+ * src: 16 bit integer to set
+ */
+#define SEED_ID_S1(dst, src) { (*(uint16_t *)&(dst)->id) = (src); (dst)->s = 1; }
+/**
+ * \brief Set the seed id to a 64 bit constant
+ * dst: seed_id_t to set to the constant
+ * src: 64 bit integer to set
+ */
+#define SEED_ID_S2(dst, src) { (*(uint64_t *)&(dst)->id) = (src); (dst)->s = 2; }
+/**
+ * \brief Set the seed id to a 128 bit constant
+ * dst: seed_id_t to set to the constant
+ * l: Lower 64 bits of the seed id to set
+ * h: Upper 64 bits of the seed id to set
+ */
+#define SEED_ID_S3(dst, l, h) { (*(uint64_t *)&(dst)->id) = (l); (*(uint64_t *)&(dst)->id[8]) = (h); (dst)->s = 3; }
+/**
+ * \brief Compare two contiki seed ids represented as seed_id_t types
+ * a: First value to compare
+ * b: Second value to compare
+ */
+#define seed_id_cmp(a, b) (memcmp((a)->id, (b)->id, sizeof(uint8_t) * 16) == 0)
+/**
+ * \brief Copy one seed_id_t into another.
+ * a: Destination
+ * b: Source
+ */
+#define seed_id_cpy(a, b) (memcpy((a), (b), sizeof(seed_id_t)))
+/**
+ * \brief Clear a seed id value to zero
+ * a: Value to clear
+ */
+#define seed_id_clr(a) (memset((a), 0, sizeof(seed_id_t)))
+/*---------------------------------------------------------------------------*/
+/* Buffered message set
+ *  This is implemented as a linked list since the majority of operations
+ *  involve finding the minimum sequence number and iterating up the list.
+ */
+struct mpl_msg {
+  uip_ip6addr_t srcipaddr;
+  uint8_t seq;
+  uint8_t data[UIP_BUFSIZE];
+  uint16_t size; /* Side of the data stored above */
+  uint8_t e; /* Expiration count for trickle timer */
+  struct mpl_msg *next; /* Next message in the set, or NULL if this is largest */
```

Comment:
any reason not to use the list module?

Reply:
I think the reason for this was that I initially didn't design it to be a linked list. It evolved that way once I realised how many loops were being used to access the elements in order. I agree that it should be modified to use the linked list library though, so I'll fix this.